### ISSUE TITLE
use both bs3 and bs4 in the same project but different modules

### ISSUE THREAD

**Issue by albertborsos** [yiisoft/yii2-bootstrap4#11]
I would like to use in my project both bs3 and bs4 widgets but in different modules.

Because each extension requires the same package (`bower-asset/bootstrap`) in different version constraints, it is impossible.

What do you think about to use `npm-asset/bootstrap` package in `yii2-bootstrap4` and ~~keep the `bower-asset/bootstrap`~~ use `twbs/bootstrap` in `yii2-bootstrap`?

**Comment by simialbi** [comment]
You could add bootstrap 3 npm package in composer.json in `require` section like this:
```json
    "require": {
        "npm-asset/bootstrap": "~3.0"
    }
```
And override paths for `yii2-bootstrap` assets in config like this
```php
    'components' => [
        'assetManager' => [
            'bundles' => [
                'yii\boostrap\BootstrapAsset' => [
                    'sourcePath' => '@npm/bootstrap/dist'
                ],
                'yii\bootstrap\BootstrapPluginAsset' => [
                    'sourcePath' => '@npm/bootstrap/dist'
                ]
            ]
        ]
    ]
```

**Comment by albertborsos** [comment]
I was inaccurate.

I would like to use `yiisoft/yii2-bootstrap4:1.0.x-dev` and `yiisoft/yii2-bootstrap:2.2.x-dev` with Yii 2.1. Now it conflicts because `yii2-bootstrap4` has a `yiisoft/yii2:~2.0.6` constraint. But this is a different thing.

This package uses `bower-asset/bootstrap` which is deprecated, so it is still an issue.

**Comment by albertborsos** [comment]
@cjtterabyte Thanks, I have this options in my composer.json

The problem is ~2.0.6 will never match with 2.1.x-dev
```
    - Can only install one of: yiisoft/yii2[2.0.x-dev, 2.1.x-dev].
    - yiisoft/yii2-bootstrap4 1.0.x-dev requires yiisoft/yii2 ~2.0.6 -> satisfiable by yiisoft/yii2[2.0.x-dev].
    - Installation request for yiisoft/yii2 2.1.x-dev -> satisfiable by yiisoft/yii2[2.1.x-dev].
```
But this is another issue #12 

This ticket's question is, how to use both `bootstrap:~3.0` and `bootstrap:~4.0` with official `yii2-bootstrap` packages.

**Comment by cjtterabyte** [comment]
@albertborsos I explain how I have everything configured:

composer.json: (yii2-app-advanced) 

```
"minimum-stability": "dev",
"prefer-stable": true,
"require": {
    "php": ">=5.4.0",
    "yiisoft/yii2": "~2.0.6",
    "cjtterabytesoft/yii2-theme-adminator-skeleton": "0.0.1.x-dev", --> BS4
    "yiisoft/yii2-swiftmailer": "~2.0.0 || ~2.1.0"
},
"require-dev": {
    "yiisoft/yii2-debug": "~2.0.0", --> BS3
    "yiisoft/yii2-gii": "~2.0.0", --> BS3
    "yiisoft/yii2-faker": "~2.0.0",
    "codeception/base": "^2.2.3",
    "codeception/verify": "~0.3.1"
},
```

composer.json:; extension BS4.

"minimum-stability": "stable",
"require": {
    "php": ">=7.1.0",
    "yiisoft/yii2": ">=2.0.14",
    "yiisoft/yii2-bootstrap4": "@dev",       
    "bower-asset/bootbox": "*",
    "bower-asset/Font-Awesome": "*",
    "bower-asset/ionicons": "*",
    "bower-asset/themify-icons": "*"
},

It works perfect for me,

Regards.

**Comment by kaspirovski** [comment]
Solution for bootstrap3 and bootstrap4: 

In `yiisoft/yii2-bootstrap4` file `composer.json`:
```
"require":
   "bower-asset/bootstrap4": "~4.0",
```

**Comment by kaspirovski** [comment]
@cjtterabyte Maybe replace in `yii2-bootstrap4/composer.json` package name for bootstrap4 to avoid conflict in `bower-asset/bootstrap` versions:

```
"require":
    "bower-asset/bootstrap4": "~4.0",
```

**Comment by schmunk42** [comment]
This might be possible for Bootstrap, since there's a separate package for `bootstrap4`, but in general we really need to find a solution for installing different versions of client-side packages.

**Comment by kwazaro** [comment]
So how to deal with this thread? Even if I reconfigure assetManager to use npm-assets for yii\bootstrap, I still can't install yii\bootstrap4 alongside yii\bootstrap (bootstrap 3)

**Comment by Thoulah** [comment]
@kwazaro What is stopping composer for you? Because I managed to do so:
```
$ composer show
[..]
yiisoft/yii2-bootstrap                2.0.0-beta         The Twitter Bootstrap extension for the Yii framework
yiisoft/yii2-bootstrap4               dev-master 35fe0eb The Twitter Bootstrap extension for the Yii framework
[..]
```

**Comment by kwazaro** [comment]
@Thoulah If I have ```"yiisoft/yii2-bootstrap": "~2.0.0-beta"``` or ```"yiisoft/yii2-bootstrap": "~2.0.0"``` in composer.json, the installation of ```yiisoft/yii2-bootstrap4``` fails with errors:
```
composer require --prefer-dist yiisoft/yii2-bootstrap4
Using version ^1.0@dev for yiisoft/yii2-bootstrap4
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - don't install yiisoft/yii2-bootstrap4 1.0.x-dev|install yiisoft/yii2-bootstrap4 dev-master
    - Installation request for yiisoft/yii2-bootstrap4 ^1.0@dev -> satisfiable by yiisoft/yii2-bootstrap4[1.0.x-dev].
    - Conclusion: remove bower-asset/bootstrap v3.3.7
    - Conclusion: don't install bower-asset/bootstrap v3.3.7
    - yiisoft/yii2-bootstrap4 dev-master requires bower-asset/bootstrap ~4.1.0 -> satisfiable by bower-asset/bootstrap[v4.1.1, v4.1.0].
    - Can only install one of: bower-asset/bootstrap[v4.1.1, v3.3.7].
    - Can only install one of: bower-asset/bootstrap[v4.1.0, v3.3.7].
    - Installation request for bower-asset/bootstrap (locked at v3.3.7) -> satisfiable by bower-asset/bootstrap[v3.3.7].


Installation failed, reverting ./composer.json to its original content.

```

**Comment by cjtterabyte** [comment]
@kwazaro I do not understand why you are going to use yii2-bootstrap for twbs 4, when there is a specific package yii2-bootstrap4

**Comment by Thoulah** [comment]
@kwazaro I have found most modules and widgets allow you to set the template from the call, making the loading of bootstrap 3 obsolete. I personally would not load bootstrap 3 & 4 css into the same app at the same time anyway, as this might have all kinds of negative side-effects.

**Comment by cjtterabyte** [comment]
https://github.com/yiisoft/yii2-bootstrap/issues/239

**Comment by samdark** [comment]
Won't be able to do that since that's asset-packagist and fxp limitation. If you need it, consider using npm directly.

**Comment by schmunk42** [comment]
@samdark I wanted to do some work on this topic, but it's blocked by https://github.com/wikimedia/composer-merge-plugin/issues/165 - if we have asset packages in `composer.json` by default we will run into huge isses when packages try to include two versions of the same package.

**Comment by samdark** [comment]
Yes. It's against Composer overall.
